### ISSUE TITLE
`requestUrl` doesn't allow unusual method on Android

### ISSUE THREAD

**Issue by fyears** [obsidianmd/obsidian-api#48]
One of [my plugin](https://github.com/remotely-save/remotely-save) enables webdav connections from Obsidian.

Webdav has some unusual http methods, like `PROPFIND`, `MKCOL`, etc. And, to avoid CORS issue, `requestUrl` is needed instead of `fetch` or `axios`. Obsidian desktop (Windows) version, **and iOS version**, do not restrict the methods in `requestUrl`, and the requests are constructed successfully and work as expected.

Unfortunately, Android's `requestUrl` seems to has some limitations, only allowing methods in `OPTIONS, GET, HEAD, POST, PUT, DELETE, TRACE, PATCH`!

Is it possible to fix this bug, and relax the restriction on Android? Thanks!

Here is the code to reproduce the issue, which works in desktop's console, but doesn't work in Android's console.

```js
requestUrl({url:'https://example.com',method:'PROPFIND'}).then(x=>x.text).then(x=>console.log(x))
```

...

![image](https://user-images.githubusercontent.com/1142836/159718794-116b9b90-1b41-49e1-864e-3b500f323e04.png)

...

![6458bead4778f748e3d1120756661a2](https://user-images.githubusercontent.com/1142836/159719141-7717a673-bd2c-4029-81e8-39ec248d88f7.png)

**Comment by lishid** [comment]
On Android, the native function is `HttpURLConnection.setRequestMethod` which documents itself as follows https://developer.android.com/reference/java/net/HttpURLConnection#setRequestMethod(java.lang.String)

I did a search and unfortunately you aren't allowed to use custom http methods: https://stackoverflow.com/questions/46101229/how-do-i-use-my-own-method-with-an-httpurlconnection-object-on-android

**Comment by lishid** [comment]
Maybe this works for your use case? It'll depend on whether webdav supports this:

https://stackoverflow.com/a/32503192

**Comment by fyears** [comment]
Well, the plugin actually let users input THEIR webdav providers' info, and I believe that NOT so many webdav provider supports `X-HTTP-Method-Override` (as well as CORS!) by default. (And I tried the hack trick with my account, found that the webdav cloud provider doesn't support `X-HTTP-Method-Override`, as expected.)

Though I totally understand that Obsidian might not have high priority or motivation to fix this "bug" (by introducing some extra dependencies like `okhttp`), is it still possible in the future? Here are two reasons I could think of:

1. `requestUrl` is designed to "replace" `fetch` in most cases, so naturally it should not has many extra restrictions.
2. Obsidian's api is designed to be cross-platform (as much as possible), and right now the desktop and iOS editions in fact support arbitrary http verbs. Only Android has a relatively inconsistent behavior. (But, please, do not add "http verb restrictions" to the desktop and iOS editions! 😂)

**Comment by lishid** [comment]
HttpRequest doesn't exist for Android. It's so confusing. I've been trying to play with the reflection based bypasses and they all seem to be for Java Java and not Java Android.

**Comment by lishid** [comment]
Ok I did some reflection wrangling and it seems to work. Do you have a server I can test with? All the test servers I've seen only supports the standard GET/POST/PUT/etc.

**Comment by fyears** [comment]
WOW!!!!!! YOU ARE THE BEST!!!!!

How to start a webdav server:

1. Download and install [rclone](https://rclone.org/downloads/), a famous open-source "command-line program to manage files on cloud storage", then, assuming you have a test folder only containing test files `/path/to/test/folder`, and you want a username `hello` and password `world`:
    ```bash
    rclone serve webdav /path/to/test/folder --user hello --pass world --addr 127.0.0.1:8080 --log-level DEBUG
    ```
    On Windows, I start a webdav serve like above, then I could run this in Obsidian successfully:
    ```js
     requestUrl({
        url: 'http://127.0.0.1:8080',
        method: 'PROPFIND',
        headers: {
            'Authorization': 'Basic aGVsbG86d29ybGQ='
        }
    }).then(x=>x.text).then(x=>console.log(x))
    ```
    An xml text should show up if everything goes well. The status code is 207 (not 200).
    
    And of course you need to adjust the username, password, address according to your needs. [More params here](https://rclone.org/commands/rclone_serve_webdav/).

2. Another hard-core way, since I am pretty sure you have nodejs env: using [`webdav-server`](https://github.com/OpenMarshal/npm-WebDAV-Server/wiki/Examples-%5Bv2%5D)

**Comment by fyears** [comment]
But please be aware that, webdav, and other http extensions, contain verbs not only `PROPFIND`. Does the reflections work for arbitrary verbs? And maybe you could pre-check the verbs, and only invoke the reflections when the method is not in the default allow list, so that the performance could be guaranteed in most cases.

**Comment by lishid** [comment]
Yeah so it should work for arbitrary words. What I'm doing is setting the method like before, but I catch the Exception that is thrown and try the reflection method. I didn't want to bother setting up a full webdav server but I did use a couple of different websites that are designed to test http requests - though none of them supported non-standard HTTP verbs, at the very least it gave me a 404 not found (which, most of the time, signals that the route with that verb was not set).

I think this should work.

**Comment by fyears** [comment]
Yeah, and maybe just trying calling https://example.com is sufficient. I call it on Windows Obsidian and get a 405 method not allowed result, as expected.

So will this fix be available in next (insider) release? (1.1.2? 1.1.3?) Thanks!

**Comment by lishid** [comment]
Actually there are a lot of problems with setting the request body on unsupported method types, even if we successfully set the method type. I guess we'll have to resort to the apache http client when using an unsupported method instead.

**Comment by lishid** [comment]
Ok I have done a few hours of source code reading and it seems like internally Android supports WebDAV methods, but only the first level API (setRequestMethod) fails. That means I can proceed with just using Reflection to set the method, without having to bring in another http library (okhttp3 made the app bundle 7mb -> 8mb).

This will be available in the next insider build of 1.2.0.

**Comment by fyears** [comment]
ok, thanks!
